### PR TITLE
[5.1][String] Scalar-alignment bug fixes.

### PR DESCRIPTION
**Pull request by milseman**
5.1 cherry-pick of https://github.com/apple/swift/pull/23834

Explanation: All String views use the same index type, with the semantics being that we interpret them as representing the nearest prior scalar-aligned position. However, we were not checking or aligning indices previously, leading to crashes in the (somewhat uncommon) situation that a non-scalar-aligned index was interchanged between sub-scalar views. This change checks and performs alignment, and allocates a known-to-be-scalar-aligned bit in String.Index’s ABI to recoup the performance back.

Scope: Swift Standard Library

Issue: <rdar://problem/51199363> and https://bugs.swift.org/browse/SR-10124

Risk: Low. There is a chance that code that was improperly exchanging indices between different Strings (not just different views) will not get the correct behavior provided by this fix, if the known-to-be-aligned bit were set from one String.

Testing: Full integration testing, benchmarking

Reviewer: @stephentyrone, @lorentey, @Catfish-Man

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test os x platform

**Comment by lorentey** [comment]
```
11:00:17 Test Suite 'Selected tests' started at 2019-06-28 11:00:03.384
11:00:17 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-28 11:00:03.385
11:00:17 Test Suite 'PackageDescription4_2LoadingTests' started at 2019-06-28 11:00:03.385
11:00:17 Test Case '-[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl]' started.
11:00:17 <unknown>:0: error: -[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
11:00:17 Test Case '-[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl]' failed (0.435 seconds).
11:00:17 Test Suite 'PackageDescription4_2LoadingTests' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
11:00:17 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
11:00:17 Test Suite 'Selected tests' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
```

**Comment by lorentey** [comment]
@swift-ci test macOS platform

**Comment by lorentey** [comment]
🤔
```
15:13:03 ("Stack dump:\n0.\tProgram arguments: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swift-macosx-x86_64/bin/swift -frontend -interpret /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/TemporaryFile.9qwtEz.swift -target x86_64-apple-macosx10.10 -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -I /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/lib/swift/pm/4_2 -swift-version 5 -package-description-version 5 -module-name main -lPackageDescription -- -fileno 42 \n0  swift                       0x000000010e6fefe5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37\n1  swift                       0x000000010e6fe028 llvm::sys::RunSignalHandlers() + 248\n2  swift                       0x000000010e6ff5d8 SignalHandler(int) + 264\n3  libsystem_platform.dylib    0x00007fff63db6b5d _sigtramp + 29\n4  libsystem_platform.dylib    0x00007ff56de43bc0 _sigtramp + 168349824\n5  libPackageDescription.dylib 0x0000000115e39bb1 $s18PackageDescription14AppleOSVersion33_EA5579823E05354E291299F0EC3450AALLPAAE6stringxSS_tcfCAA17SupportedPlatformV03MacD0V_Tg5Tf4xd_n + 161\n6  libPackageDescription.dylib 0x0000000115e33cf2 globalinit_33_EA5579823E05354E291299F0EC3450AA_func10 + 82\n7  libdispatch.dylib           0x00007fff63b7e63d _dispatch_client_callout + 8\n8  libdispatch.dylib           0x00007fff63b7fd4c _dispatch_once_callout + 20\n9  libswiftCore.dylib          0x00007fff636047e9 swift_once + 25\n10 libPackageDescription.dylib 0x0000000115e33d31 $s18PackageDescription17SupportedPlatformV12MacOSVersionV6v10_13AEvau + 33\n11 libPackageDescription.dylib 0x0000000115eb5079 $s18PackageDescription17SupportedPlatformV12MacOSVersionV6v10_13AEvau + 529257\n12 swift                       0x000000010ab98258 llvm::MCJIT::runFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 456\n13 swift                       0x000000010ab9b3c1 llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, char const* const*) + 1313\n14 swift                       0x000000010ab8daa1 swift::RunImmediately(swift::CompilerInstance&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, swift::IRGenOptions&, swift::SILOptions const&) + 3505\n15 swift                       0x000000010ab7e075 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 13653\n16 swift                       0x000000010ab79b52 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2978\n17 swift                       0x000000010ab22a28 main + 696\n18 libdyld.dylib               0x00007fff63bcb3d5 start + 1", diagnosticFile: nil)
15:13:03 Test Suite 'Selected tests' started at 2019-06-28 15:12:46.883
15:13:03 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-28 15:12:46.883
15:13:03 Test Suite 'PackageDescription5LoadingTests' started at 2019-06-28 15:12:46.884
15:13:03 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' started.
15:13:03 /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swiftpm/Tests/PackageLoadingTests/PD5LoadingTests.swift:166: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed
15:13:03 <unknown>:0: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
15:13:03 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' failed (7.880 seconds).
15:13:03 Test Suite 'PackageDescription5LoadingTests' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.880) seconds
15:13:03 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.880) seconds
15:13:03 Test Suite 'Selected tests' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.881) seconds
```

**Comment by lorentey** [comment]
@swift-ci test macOS platform

**Comment by lorentey** [comment]
```
02:02:45 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-29 02:02:32.009
02:02:45 Test Suite 'PackageDescription5LoadingTests' started at 2019-06-29 02:02:32.009
02:02:45 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' started.
02:02:45 /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swiftpm/Tests/PackageLoadingTests/PD5LoadingTests.swift:166: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed
02:02:45 <unknown>:0: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
02:02:45 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' failed (6.399 seconds).
02:02:45 Test Suite 'PackageDescription5LoadingTests' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
02:02:45 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
02:02:45 Test Suite 'Selected tests' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
```

**Comment by milseman** [comment]
Also cherry-picking https://github.com/apple/swift/pull/25948

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

File: stdlib/public/core/Substring.swift [hunk 1]
```diff
@@ -97,9 +97,15 @@ public struct Substring {
   @usableFromInline
   internal var _slice: Slice<String>
 
-  @inlinable @inline(__always)
```
Comment (stephentyrone) [inline stdlib/public/core/Substring.swift]: Expected?

**Comment by milseman** [comment]
@swift-ci please nominate

File: stdlib/public/core/UnicodeHelpers.swift [hunk 2]
```diff
@@ -140,10 +140,12 @@ internal func _continuationPayload(_ x: UInt8) -> UInt32 {
   return UInt32(x & 0x3F)
 }
 
-@inlinable @inline(__always)
```
Comment (stephentyrone) [inline stdlib/public/core/UnicodeHelpers.swift]: Also this one, to double-check?

File: test/stdlib/StringIndex.swift [hunk 3]
```diff
@@ -232,21 +227,238 @@ StringIndexTests.test("String.Index(_:within) / Range<String.Index>(_:in:)") {
...
+StringIndexTests.test("Exhaustive Index Interchange") {
+  // Exhaustively test aspects of string index interchange
+  func testInterchange(
+    _ str: String,
+    stackTrace: SourceLocStack = SourceLocStack(),
+    showFrame: Bool = true,
+    file: String = #file,
+    line: UInt = #line
+  ) {
+    guard #available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) else {
```
Comment (stephentyrone) [inline test/stdlib/StringIndex.swift]: Can this be macOS 10.15, etc now?

**Comment by milseman** [comment]
@swift-ci nominate

**Review by stephentyrone — APPROVED** [review]
Generally LGTM. Flagged a couple changes to make sure they're intended (or more specifically, that this hasn't picked up some other commits).

File: stdlib/public/core/Substring.swift (hunk 1 above)
Comment (milseman) [reply stdlib/public/core/Substring.swift]: Yup, this is no longer trivial now and we don't want to force the optimizer's hand. The optimizer's probably better off not inlining this call and inlining the `scalarAlign` calls.

File: stdlib/public/core/UnicodeHelpers.swift (hunk 2 above)
Comment (milseman) [reply stdlib/public/core/UnicodeHelpers.swift]: Yeah, the inline-me fast-path has switched to the bit-check.

File: test/stdlib/StringIndex.swift (hunk 3 above)
Comment (milseman) [reply test/stdlib/StringIndex.swift]: I'm matching what's currently there. I had also tried to update all of String's versions in another commit, but hit a snag during testing (don't recall what it was, probably I wasn't running the beta). I want to get this in for 5.1 in time, so I just followed what we already have throughout the source.

**Review by Catfish-Man — APPROVED** [review]
There's a lot here to digest and I'd need to step through some of it in a debugger to be thorough, but reading through it it looks fine.

**Pull request merged**